### ISSUE TITLE
ZIO.foreachParN_ is causing JVM StackOverflow

### ISSUE THREAD

**Issue by abomm** [zio/zio#1499]
I am getting a java.lang.StackOverflowError when using ZIO.foreachParN_ or ZIO.foreachPar_.

Please run this test code, on ZIO 1.0.0-RC11-1
If you run the main with args of Array("10000", "1","1") which means build a map of 10K keys and random elements, it generates the error.

https://gist.github.com/abomm/6e3420e93958119fe312e8e5b7dbdee9

Error:
```bash
java.lang.StackOverflowError
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:212)
	at scala.collection.mutable.HashMap$$anon$3.next(HashMap.scala:165)
	at zio.ZIOFunctions.loop$4(ZIO.scala:1841)
	at zio.ZIOFunctions.loop$4(ZIO.scala:1841)
	at zio.ZIOFunctions.loop$4(ZIO.scala:1841)
```

The offending ZIO code looks something like this:
```scala
        _ ← ZIO.foreachPar_(m.keys) { k ⇒
             UIO {
               m.update(k, m(k).sorted)
               m
             }
           }

```
When this map has >10,000 keys we get blow the stack.

**Comment by jdegoes** [comment]
We will fix this for 1.0!

**Comment by LGLO** [comment]
I'd like to take this!
I'll try to come with an idea tomorrow in the morning and discuss it with someone bright.
If doesn't go well I'll release by 3pm tomorrow.

**Comment by LGLO** [comment]
OFC if someone else wants this but was shy to announce, ping me and issue is yours :D

**Comment by dzajkowski** [comment]
just a couple of notes:
- it's not trivial to replicate, I needed to change .jvmopts: `-Xss1M`
- looks like suspending the loop fixes the failing case:
```scala
  final def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, _]): ZIO[R, E, Unit] =
    ZIO
      .succeed(as.iterator)
      .flatMap { i =>
        def loop(a: A): ZIO[R, E, Unit] =
          if (i.hasNext) f(a).zipWithPar(ZIO.effectSuspendTotal(loop(i.next)))((_, _) => ())
          else f(a).unit
        if (i.hasNext) loop(i.next)
        else ZIO.unit
      }
      .refailWithTrace
```

**Comment by LGLO** [comment]
@hanny24 is taking care of it as a part of #1182

**Comment by eamelink** [comment]
Easiest way to replicate: 

    ZIO.foreachPar_(1 to 100000)(n => ZIO.unit)

Crashes with a StackOverflowError

**Comment by evis** [comment]
ZManaged is affected by the same bug.

I will take this one.

**Comment by evis** [comment]
Resolved in #2590. I guess, we may close this issue.

**Comment by adamgfraser** [comment]
Thanks for the reminder!
